net/bonding/files/lib/netifd/proto/bonding.sh

```diff
@@ -0,0 +1,237 @@
+#!/bin/sh
+#
+# Copyright (C) 2018 TDT AG <development@tdt.de>
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See https://www.gnu.org/licenses/gpl-2.0.txt for more information.
+#
+
+. /lib/functions.sh
+. ../netifd-proto.sh
+
+init_proto "$@"
+
+INCLUDE_ONLY=1
+
+BONDING_MASTERS="/sys/class/net/bonding_masters"
+
+proto_bonding_init_config() {
+	no_device=1
+	available=1
+
+	proto_config_add_string "dhcp"
+	proto_config_add_string "ifname"
+
+	proto_config_add_string "ipaddr"
+	proto_config_add_string "netmask"
+
+	proto_config_add_string "bonding_policy"
+	proto_config_add_string "link_monitoring"
+	proto_config_add_string "slaves"
+	proto_config_add_string "all_slaves_active"
+
+	proto_config_add_string "min_links"
+	proto_config_add_string "ad_actor_sys_prio"
+	proto_config_add_string "ad_actor_system"
+	proto_config_add_string "ad_select"
+	proto_config_add_string "lacp_rate"
+	proto_config_add_string "packets_per_slave"
+	proto_config_add_string "xmit_hash_policy"
+	proto_config_add_string "primary"
+	proto_config_add_string "primary_reselect"
+	proto_config_add_string "lp_interval"
+	proto_config_add_string "tlb_dynamic_lb"
+	proto_config_add_string "resend_igmp"
+	proto_config_add_string "fail_over_mac"
+	proto_config_add_string "num_grat_arp__num_unsol_na"
+
+	proto_config_add_string "arp_interval"
+	proto_config_add_string "arp_ip_target"
+	proto_config_add_string "arp_all_targets"
+	proto_config_add_string "arp_validate"
+
+	proto_config_add_string "miimon"
+	proto_config_add_string "downdelay"
+	proto_config_add_string "updelay"
+	proto_config_add_string "use_carrier"
+}
+
+proto_bonding_setup() {
+	local cfg="$1"
+	local link="bonding-$cfg"
+
+	# Check for loaded kernel bonding driver (/sys/class/net/bonding_masters exists)
+	[ -f "$BONDING_MASTERS" ] || {
+		echo "$cfg" "setup: bonding_masters does not exist in sysfs (kernel module not loaded?)"
+		proto_notify_error "$cfg" "setup: bonding_masters does not exist in sysfs (kernel module not loaded?)"
+		proto_block_restart "$cfg"
+		return
+	}
+
+	# Add bonding interface to system
+	echo "+$link" > "$BONDING_MASTERS"
+
+	# Set bonding policy (with corresponding parameters)
+	local bonding_policy
+	json_get_vars bonding_policy
+
+	[ -n "$bonding_policy" ] && echo "$bonding_policy" > /sys/class/net/"$link"/bonding/mode
```

Comment:
Maybe you should move this into each case switch below? This would allow for some validation of the mode value.

Reply:
Modified the code, but I'm not sure if I got you right.
Please review the changed part again and tell me if this is what you meant